### PR TITLE
Feature/adding program options state writer

### PR DESCRIPTION
**Pull request by gsps1**
https://issues.cask.co/browse/CDAP-13352
This covers Part-1 of changes required for CDAP-13352.
Changes
1) API change in ProgramStateWriter lifecycle methods to pass along ProgramOptions, required for  recording relevant additional details about the program status. 
2) Heartbeat thread to send heartbeat type notification while program is running based on configured interval, default : 30 mins

### CODE REVIEW

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/program/MessagingProgramStatePublisher.java [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+import com.google.gson.GsonBuilder;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+import java.util.Map;
+import java.util.concurrent.TimeUnit;
+
+/**
+ * publishes program state and heart beat messages through the messaging service
```
Comment (ecfm) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/program/MessagingProgramStatePublisher.java]: Capitalize class javadoc, heartbeat is a single word

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/program/ProgramRunHeartbeat.java [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+ */
+public class ProgramRunHeartbeat implements Runnable {
+  private static final Logger LOG = LoggerFactory.getLogger(ProgramRunHeartbeat.class);
+
+  private final ProgramStatePublisher programStatePublisher;
+  private final Map<String, String> properties;
+  private final String programRunId;
+
+
+  ProgramRunHeartbeat(ProgramStatePublisher messagingProgramStatePublisher, Map<String, String> properties) {
```
Comment (ecfm) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/program/ProgramRunHeartbeat.java]: `messagingProgramStatePublisher ` is not necessarily a messaging publisher

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/program/ProgramStatePublisher.java [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+import java.util.Map;
+
+/**
+ * Publishing program state messages and heartbeats
+ */
+public interface ProgramStatePublisher {
+
+  /**
+   * Publish message which is identified by notificationType and the properties.
+   * @param notificationType type of the notification of the message
```
Comment (ecfm) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/program/ProgramStatePublisher.java]: add new line above

**Comment by gsps1** [comment]
i am fixing issues based on testing in the cluster. will push the changes after completing the testing.

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/program/ProgramRunHeartbeat.java (hunk 2 above)
Comment (gsps1) [reply cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/program/ProgramRunHeartbeat.java]: its a program state writer that uses messaging service. if you want it  to be renamed, suggest a better name. thanks

**Comment by gsps1** [comment]
build running : https://builds.cask.co/browse/CDAP-DUT6504-1

File: cdap-app-fabric/src/main/java/co/cask/cdap/app/runtime/ProgramStateWriter.java [hunk 4]
```diff
@@ -44,42 +45,48 @@ void start(ProgramRunId programRunId, ProgramOptions programOptions, @Nullable S
    *
    * @param programRunId the id of the program run
    * @param twillRunId the run id of the twill application
+   * @param programOptions options passed to the program run
    */
-  void running(ProgramRunId programRunId, @Nullable String twillRunId);
+  void running(ProgramRunId programRunId, @Nullable String twillRunId, ProgramOptions programOptions);
```
Comment (albertshau) [inline cdap-app-fabric/src/main/java/co/cask/cdap/app/runtime/ProgramStateWriter.java]: why are these changes needed? When would the program options ever change during a run?

File: cdap-app-fabric/src/main/java/co/cask/cdap/app/runtime/AbstractProgramRuntimeService.java [hunk 5]
```diff
@@ -146,14 +145,13 @@ public final RuntimeInfo run(ProgramDescriptor programDescriptor, ProgramOptions
...
-
       RuntimeInfo runtimeInfo = createRuntimeInfo(runner.run(executableProgram, optionsWithPlugins), programId,
-                                                  cleanUpTask);
+                                                  cleanUpTask, runtimeProgramOptions);
       monitorProgram(runtimeInfo, cleanUpTask);
       return runtimeInfo;
     } catch (Exception e) {
       // Set the program state to an error when an exception is thrown
-      programStateWriter.error(programId.run(runId), e);
+      programStateWriter.error(programId.run(runId), e, runtimeProgramOptions);
```
Comment (albertshau) [inline cdap-app-fabric/src/main/java/co/cask/cdap/app/runtime/AbstractProgramRuntimeService.java]: you mentioned this is one of two places an error is written. Please remove the duplicates. I don't think this is the right place to write the error or to error log, as the caller needs to handle the exception anyway, and the logical thing for the caller to do is write an error and log an error.

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/program/MessagingProgramStateWriter.java [hunk 6]
```diff
@@ -78,106 +92,102 @@ public void start(ProgramRunId programRunId, ProgramOptions programOptions, @Nul
...
+  public void suspend(ProgramRunId programRunId, ProgramOptions programOptions) {
+    programStatePublisher.publish(Notification.Type.PROGRAM_STATUS,
+                                  ImmutableMap.<String, String>builder()
+                                    .put(ProgramOptionConstants.PROGRAM_RUN_ID, GSON.toJson(programRunId))
+                                    .put(ProgramOptionConstants.SUSPEND_TIME,
+                                         String.valueOf(System.currentTimeMillis()))
+                                    .put(ProgramOptionConstants.PROGRAM_STATUS, ProgramRunStatus.SUSPENDED.name())
+                                    .put(ProgramOptionConstants.PROGRAM_OPTIONS, GSON.toJson(programOptions)).build()
     );
+    if (scheduledExecutorService != null) {
```
Comment (albertshau) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/program/MessagingProgramStateWriter.java]: you should shutdown before you publish, otherwise you can still sent other messages after publishing

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/program/ProgramRunHeartbeat.java [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.HashMap;
+import java.util.Map;
+
+/**
+ * Heartbeat thread to publish heart beat messages with notification type heart_beat
+ */
+public class ProgramRunHeartbeat implements Runnable {
```
Comment (albertshau) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/program/ProgramRunHeartbeat.java]: you don't need this class, you can inline it like:
```
() -> {  
   properties.put(ProgramOptionConstants.HEART_BEAT_TIME, String.valueOf(System.currentTimeMillis()));
   // publish as heart_beat type, so it can be handled appropriately at receiver
   programStatePublisher.publish(Notification.Type.HEART_BEAT, properties);
}
```

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/program/ProgramRunHeartbeat.java [hunk 8]
```diff
@@ -0,0 +1,53 @@
...
+
+  private final ProgramStatePublisher programStatePublisher;
+  private final Map<String, String> properties;
+  private final String programRunId;
+
+  ProgramRunHeartbeat(ProgramStatePublisher messagingProgramStatePublisher, Map<String, String> properties) {
+    this.programStatePublisher = messagingProgramStatePublisher;
+    this.properties = new HashMap<>(properties);
+    // could be called from RESUMING state, so update status to running for heartbeat
+    this.properties.put(ProgramOptionConstants.PROGRAM_STATUS, ProgramRunStatus.RUNNING.name());
```
Comment (albertshau) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/program/ProgramRunHeartbeat.java]: This won't be an issue if you inline the class, but you don't want to make assumptions like this when creating a separate class, as anybody can create this class. At a quick glance, it looks like it just publishes a notification with the given publisher and given properties. The fact that it modifies the properties is hidden which can easily lead to misuse

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/program/MessagingProgramStateWriter.java [hunk 9]
```diff
@@ -53,16 +53,30 @@
...
-  private final TopicId topicId;
-  private final RetryStrategy retryStrategy;
+  private static final Gson GSON =
+    ApplicationSpecificationAdapter.addTypeAdapters(new GsonBuilder())
+      .registerTypeAdapter(Arguments.class, new ArgumentsCodec())
+      .registerTypeAdapter(ProgramOptions.class, new ProgramOptionsCodec()).create();
+
+  private final ProgramStatePublisher programStatePublisher;
+
+  private ScheduledExecutorService scheduledExecutorService;
```
Comment (albertshau) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/program/MessagingProgramStateWriter.java]: This scheduling is at the wrong level, it should not be in this class. There is nothing in the StateWriter contract that forces you to use a separate instance per program run, but this implementation assumes there is one writer per run. For example, the state writer is also used in the CDAP master, where this logic will be wrong.

The twill application should have a scheduled service that uses the state writer, not the other way around.

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/services/ProgramNotificationSubscriberService.java [hunk 10]
```diff
@@ -357,6 +358,9 @@ private void handleProgramEvent(ProgramRunId programRunId, ProgramRunStatus prog
           systemArgs.put(ProgramOptionConstants.CLUSTER_KEY_INFO,
                          properties.get(ProgramOptionConstants.CLUSTER_KEY_INFO));
         }
+        systemArgs.put(ProgramOptionConstants.ARTIFACT_ID,
```
Comment (albertshau) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/services/ProgramNotificationSubscriberService.java]: if these things are needed in the system args, they should be placed there at the start of the run (the runInternal method), not later on in the lifecycle. This probably goes for whatever else is being updated within the AbstractProgramRuntimeService. Is there a reason they can't be set at the very beginning?

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/services/RunRecordCorrectorService.java [hunk 11]
```diff
@@ -146,6 +153,23 @@ void fixRunRecords() {
     return fixedPrograms;
   }
 
+  private ProgramOptions constructAndGetProgramOptions(RunRecordMeta record) {
```
Comment (albertshau) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/services/RunRecordCorrectorService.java]: the run record corrector should not need to know how to create ProgramOptions from a run record. Also, why do we need to add things to the system properties instead of using the existing properties? Wouldn't this only apply to records that started before 5.0.0 but where corrected after 5.0.0?

**Review by albertshau — COMMENTED** [review]
next time please be more descriptive in the commit messages and the jira so that somebody without any context can understand why the change is being made. I had no context for this change so it took half an hour to figure out what the general purpose is, and I still don't know how the heartbeat is going to be used

**Comment by gsps1** [comment]
@albertshau thanks for the review. Sorry about the missing context. I didn't realize that the design link in JIRA wasn't the correct one. I had documented the use case and design here https://wiki.cask.co/display/CE/Active+program+runs+heartbeat. 

The purpose of adding heartbeat is to get actively running/completed program over a time range efficiently. which is very inefficient if we have to use the app meta store to read run records, due to the rowkey format and we have to do full table scan for any queries for active runs in a time-range.

By using heartbeat thread using configurable interval, we will be writing records for running runs and other program status to a heartbeat table, so we will be able to retrieve results efficiently and only look into records in time range of the query we are interested in and not do a full table scan.

We are adding programOptions to other lifecycle states, as we need information such as artifactId, principal which are not part of programRunId.

File: cdap-app-fabric/src/main/java/co/cask/cdap/app/runtime/ProgramStateWriter.java (hunk 4 above)
Comment (gsps1) [reply cdap-app-fabric/src/main/java/co/cask/cdap/app/runtime/ProgramStateWriter.java]: it would not change, but its currently only available in start method (called from master), which wont be available for methods such as running, etc as they are called from Twill AM container. thats why we pass programOptions to these methods.

File: cdap-app-fabric/src/main/java/co/cask/cdap/app/runtime/AbstractProgramRuntimeService.java (hunk 5 above)
Comment (gsps1) [reply cdap-app-fabric/src/main/java/co/cask/cdap/app/runtime/AbstractProgramRuntimeService.java]: yes, i will change the logic to throw the error.

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/services/RunRecordCorrectorService.java (hunk 11 above)
Comment (gsps1) [reply cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/services/RunRecordCorrectorService.java]: umm, are you saying system properties will include artifactId already and i am redoing a put?  i can verify if that is the case. please clarify.

**Comment by albertshau** [comment]
For the other lifecycle states, all that information is set once when it records the very first run state (provisioning). For subsequent states, the information is already there in the existing run record, so it is just carried forward to the new run record. Is there a reason we can't do the same for that new table?

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/services/RunRecordCorrectorService.java (hunk 11 above)
Comment (albertshau) [reply cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/services/RunRecordCorrectorService.java]: ProgramOptions do not change throughout the course of the run. They are only set once at the start. So if they are set in the beginning, there is no need to re-set them here.

**Comment by gsps1** [comment]
had discussion with @chtyim and @albertshau, decided to use run-record table as source of truth and remove program options added to the API. for heart beat messages, we will just pass run-id and heartbeat time as identifier. will update the PR with these changes.

**Comment by gsps1** [comment]
@albertshau updated the PR as per our discussion. please take a look again, thanks.

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/program/ProgramStateWriterWithHeartBeat.java [hunk 12]
```diff
@@ -0,0 +1,155 @@
...
+      scheduledExecutorService =
+        Executors.newSingleThreadScheduledExecutor(Threads.createDaemonThreadFactory("program-heart-beat"));
+      scheduledExecutorService.scheduleAtFixedRate(
+        () -> {
+          Map<String, String> properties = new HashMap<>();
+          properties.put(ProgramOptionConstants.PROGRAM_RUN_ID, GSON.toJson(programRunId));
+          properties.put(ProgramOptionConstants.HEART_BEAT_TIME, String.valueOf(System.currentTimeMillis()));
+          // publish as heart_beat type, so it can be handled appropriately at receiver
+          messagingProgramStatePublisher.publish(Notification.Type.HEART_BEAT, properties);
+          LOG.trace("Recording heartbeat for program {}", programRunId);
```
Comment (albertshau) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/program/ProgramStateWriterWithHeartBeat.java]: more accurate to say 'Sent heartbeat ...'

File: cdap-app-fabric/src/test/java/co/cask/cdap/internal/app/program/ProgramStateWriterWithHeartBeatTest.java [hunk 13]
```diff
@@ -0,0 +1,112 @@
...
+
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.TimeoutException;
+
+public class ProgramStateWriterWithHeartBeatTest {
+  class MockProgramStatePublisher implements ProgramStatePublisher {
```
Comment (albertshau) [inline cdap-app-fabric/src/test/java/co/cask/cdap/internal/app/program/ProgramStateWriterWithHeartBeatTest.java]: can make this private static

File: cdap-app-fabric/src/test/java/co/cask/cdap/internal/app/program/ProgramStateWriterWithHeartBeatTest.java [hunk 14]
```diff
@@ -0,0 +1,112 @@
...
+    Tasks.waitFor(true , () -> ((MockProgramStatePublisher) programStatePublisher).getHeartBeatCount() > 1,
+                  10, TimeUnit.SECONDS, "Didn't receive expected heartbeat after 10 seconds");
+
+    // make sure suspending program suspended the heartbeat thread
+    programStateWriterWithHeartBeat.suspend();
+    Tasks.waitFor(false , () -> programStateWriterWithHeartBeat.isHeartBeatThreadAlive(),
+                  5, TimeUnit.SECONDS, "Heartbeat thread did not stop after 5 seconds");
+    long heartBeatAfterSuspend = ((MockProgramStatePublisher) programStatePublisher).getHeartBeatCount();
+
+    // resume the program and make sure that the hear beat messages goes up after resuming program
```
Comment (albertshau) [inline cdap-app-fabric/src/test/java/co/cask/cdap/internal/app/program/ProgramStateWriterWithHeartBeatTest.java]: typo: 'hear' -> 'heart'

File: cdap-proto/src/main/java/co/cask/cdap/proto/Notification.java [hunk 15]
```diff
@@ -46,7 +46,8 @@
     TIME,
     STREAM_SIZE,
     PARTITION,
-    PROGRAM_STATUS
+    PROGRAM_STATUS,
+    HEART_BEAT
```
Comment (albertshau) [inline cdap-proto/src/main/java/co/cask/cdap/proto/Notification.java]: think it's better to name it PROGRAM_HEART_BEAT, just in case we ever have other heartbeats.

**Review by albertshau — COMMENTED** [review]
only minor comments. Good job, this is now much cleaner and easier to review.

**Comment by gsps1** [comment]
thanks @albertshau, addressed the final comments.

**Review by albertshau — APPROVED** [review]
lgtm

**Comment by gsps1** [comment]
build passed : https://builds.cask.co/browse/CDAP-DUT6504-3 merging.
